### ISSUE TITLE
Security considerations

### ISSUE THREAD

**Issue by lmeyerov** [stedolan/jq#1361]
We're considering allowing user-defined transforms specified in jq, but it has been difficult for us to ascertain how safe it is from code injection & privilege escalation, by design & in practice. As far as I can tell, while jq supports some user-defined functions, they are largely sandboxed. The exception is the module imports, which would need some thing for restricting.

Can anyone clarify the security model? May help to link the answer as part of the README.

**Comment by nicowilliams** [comment]
Excellent question.  So far we've held back from adding the sorts of builtins that might make it easier to break out of the jq sandbox.

Real or potential issues to be concerned with:

 - CPU consumption -- jq is Turing complete (but single-threaded)
 - Row-hammer attacks -- there's no reason to think that they are not possible to code in jq
 - Potential security bugs in jq -- we've not yet applied AFL, but others have and we know that there exist JSON parser bugs
 - Potential security bugs in bison where the jq program could source itself can exploit them
 - It is possible to read arbitrary files down from the current directory using `include` and `import` directives -- this is a big deal if you don't trust the jq code you're running
 - We do want to add more I/O builtins in the future -- you can't count on jq being sandboxed forever

Now, it is my intention to require the use of a command-line switch to enable future I/O builtins that could allow reading or writing files, or executing external programs.  But still, there's no way to guarantee that's so.

I think it's probably desirable to add a command-line option to disallow the main program from including files or modules, or more generally a sandbox option.

**Comment by nicowilliams** [comment]
See also #1215, #660, and others.

**Comment by lmeyerov** [comment]
Ah, excellent. Thanks! 

Out of those, the include and import statements seem concerning under most threat models (including ours.) +1 for sandboxed mode flag for this and future features :) 

For the interim, would it be sufficient for us to do a blacklist for "include" and "import"?

**Comment by nicowilliams** [comment]
@lmeyerov Yes.  You might even edit `src/parser.y` and remove the `include`/`import` functionality.

Thinking about how to disable `include`/`import`... we could just shove that into `struct locfile`, or add another parse param for this, and the in the rules for `include`/`import` simply error out if they are disabled.

**Comment by nicowilliams** [comment]
There's a lot of potential sandboxing options:

 - allow importing modules, but not data
 - allow importing modules, but not data, and don't allow modules to import data either
 - don't allow importing anything

Eventually we'll have fairly capable I/O builtins.  We might have a need for many sandboxing options.

Obviously, @lmeyerov, the simplest sandbox will be to `chroot` jq...

**Comment by nicowilliams** [comment]
Actually, that may well be the best thing we could do: have a "chroot" option that disallows I/O to/from files outside the sandbox.  On Unix-like systems jq would then literally call `chroot(2)` and be done, but on Windows it might attempt to implement a chroot-like sandbox for the jq program in user-land.

**Comment by lmeyerov** [comment]
It may help, as simple policies, to consider two of the most basic security properties: data confidentiality & integrity. I'd expect "--sandbox" to provide both. Integrity doesn't seem to be an issue, but confidentiality would seem to require disallow both module & data import, as who knows what's on the system. (Also, I'd assuming the attacker can control the data & transform parameters, but not the rest .)

FWIW, we're using JQ via https://github.com/sanack/node-jq, so a `--sandbox` would be easy to use from there.

**Comment by jemc** [comment]
If a PR was submitted to add a `--sandbox` option that prevented the use of `include`/`import` (and would be expected to also prevent any future I/O primitives that could break the sandbox), would that PR be accepted (assuming any review comments about the details of the implementation were properly addressed)?

Any notes from the maintainers about acceptance criteria for somebody who wants to do that PR?

**Comment by jemc** [comment]
Okay, this was a simple enough change that I decided to go ahead and [create a pull request to add the `--sandbox` flag](https://github.com/jqlang/jq/pull/3092).

Hopefully that can mitigate at least this category of risk (including/importing arbitrary files/data from the file system), if the PR is accepted.
